rtgang++/rtgang++-v6.5.patch

```diff
@@ -0,0 +1,1512 @@
+  */
+ SCHED_FEAT(GENTLE_FAIR_SLEEPERS, true)
+ 
++/*
++ * Enable real-time gang scheduling framework (RT-Gang). RT-Gang allows
++ * execution of a single (multi-threaded) real-time task (i.e., gang) at any
++ * giving time across all system cores.
++ * NOTE: This feature is disabled by default.
++ */
++SCHED_FEAT(RT_GANG_LOCK, false)
++
+ /*
+  * Place new tasks ahead so that they do not starve already running
+  * tasks
+diff --git a/kernel/sched/rt.c b/kernel/sched/rt.c
+index 590abc1a0..505b73688 100644
+--- a/kernel/sched/rt.c
++++ b/kernel/sched/rt.c
+@@ -4,10 +4,202 @@
+  * policies)
+  */
+ 
++#include <linux/debugfs.h>
++#include <linux/uaccess.h>
++/* RT-Gang Definitions */
++rt_gang_lock_t glocks[NR_CPUS];
++rt_gang_lock_t*	rt_glock = &glocks[0];
++int core_to_gang[NR_CPUS];
++int be_mem_threshold = SYS_MAX_LLC_EVENTS;
++EXPORT_SYMBOL(be_mem_threshold);
++
++static struct dentry *rtgang_dir;
++
++static ssize_t rtgang_coremap_write(struct file *filp,
++				    const char __user *ubuf,
++				    size_t cnt, loff_t *ppos)
++{
++	int core, map;
++	int BUF_SIZE = 256;
++	char buf[256];
++	char *p = buf;
++	
++	// Only allow core mapping changes when RT-Gang is disabled
++	if(sched_feat(RT_GANG_LOCK))
++	{
++		pr_err("ERR: Core mappings can only be changed under NO_RT_GANG_LOCK\n");
++		return cnt;
++	}
++
++	if (copy_from_user(&buf, ubuf, (cnt > BUF_SIZE) ? BUF_SIZE: cnt) != 0) 
++		return 0;
++	
++	// Get and validate the core and new mapping
++	sscanf(p, "%d %d", &core, &map);
++	if(core < 0 || core >= NR_CPUS)
++	{
++		pr_err("ERR: Core %i is out of bounds\n", core);
++		return cnt;
++	}
++	else if(map < 0 || map >= NR_CPUS)
++	{
++		pr_err("ERR: Mapping given for core %i is out of bounds\n", map);
++		return cnt;
++	}
++	
++	// Only change the mapping if both inputs are valid
++	core_to_gang[core] = map;
++	
++	return cnt;
++}
++
++static int rtgang_coremap_show(struct seq_file *m, void *v)
++{
++	int i;
++	seq_printf(m, "Core to Gang Mappings\n");
++	for(i = 0; i < NR_CPUS; i++)
++	{
++		seq_printf(m, "%i --> %i\n", i, core_to_gang[i]);
++	}
++	return 0;
++}
++
++static int rtgang_coremap_open(struct inode *inode, struct file *filp)
++{
++	return single_open(filp, rtgang_coremap_show, NULL);
++}
++
++static const struct file_operations rtgang_coremap_fops = {
++	.open		= rtgang_coremap_open,
++	.write      = rtgang_coremap_write,
++	.read		= seq_read,
++	.llseek		= seq_lseek,
++	.release	= single_release,
++};
++
++static int rtgang_init_debugfs(void)
++{
++
++	rtgang_dir = debugfs_create_dir("rtgang", NULL);
++	if(!rtgang_dir)
++	{
++		goto fail;
++	}
++	if(!debugfs_create_file("coremap", 0444, rtgang_dir, NULL,
++			    &rtgang_coremap_fops))
++	{
++		goto fail;
++	}
++	return 0;
++	
++fail:
++	debugfs_remove_recursive(rtgang_dir);
++ 	return -ENOMEM;
++}
++
++late_initcall(rtgang_init_debugfs);
++
++/* End of RT-Gang Definitions */
++
++
++
+ int sched_rr_timeslice = RR_TIMESLICE;
+ /* More than 4 hours if BW_SHIFT equals 20. */
```

Comment:
is this comment still valid?

Reply:
It looks like the comment is still valid. It is part of the original code base.

Here is 6.5 unpatched 
<img width="428" alt="6 5clean" src="https://github.com/CSL-KU/ArmArHudChallenge/assets/82488869/a4a5356e-737b-4ac4-8759-839e596e6ae9">

Compared to 6.5 patch
<img width="340" alt="6 5patched" src="https://github.com/CSL-KU/ArmArHudChallenge/assets/82488869/938f5254-3b68-4123-933d-edd5fb9ecce3">


There was a slight change to this segment since 5.15. Here is what this segment looked like then
<img width="400" alt="5 15" src="https://github.com/CSL-KU/ArmArHudChallenge/assets/82488869/f396c5c8-3932-4b4b-aa27-adf72c1c3ddc">